### PR TITLE
E2e to verify lighthouse return cluster-ip from healthy service

### PR DESCRIPTION
**Pull request by aswinsuryan**
Fixes: submariner-io/submariner#1041

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

### CODE REVIEW

File: test/e2e/discovery/service_discovery.go [hunk 1]
```diff
@@ -81,6 +81,12 @@ var _ = Describe("[discovery] Test Service Discovery Across Clusters", func() {
 			RunServiceDiscoveryRoundRobinTest(f)
 		})
 	})
+
+	When("Only one of the clusters with service is healthy", func() {
+		It("should not resolve the service on the unhealthy cluster", func() {
```
Comment (tpantelis) [inline test/e2e/discovery/service_discovery.go:86]:
```suggestion
	When("only one of the clusters with a service is healthy", func() {
		It("should not resolve the service", func() {
```

File: test/e2e/framework/framework.go [hunk 2]
```diff
@@ -416,3 +435,40 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
 
 	return result
 }
+
+func (f *Framework) GetHealthCheckIpInfo(cluster framework.ClusterIndex) (endpointName, healthCheckIP string) {
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:439]:
```suggestion
func (f *Framework) GetHealthCheckIPInfo(cluster framework.ClusterIndex) (endpointName, healthCheckIP string) {
```

File: test/e2e/framework/framework.go [hunk 3]
```diff
@@ -416,3 +435,40 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+
+		if strings.Contains(endpoint.GetName(), framework.TestContext.ClusterIDs[cluster]) {
+			endpointName = endpoint.GetName()
+			Expect(endpointName).NotTo(BeNil())
+
+			spec, found, err := unstructured.NestedMap(endpoint.Object, "spec")
+
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(found).NotTo(BeFalse())
+			healthCheckIP, found, err = unstructured.NestedString(spec, "healthCheckIP")
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:455]:
```suggestion
			healthCheckIP, found, err = unstructured.NestedString(endpoint.Object, "spec", "healthCheckIP")
```

File: test/e2e/framework/framework.go [hunk 4]
```diff
@@ -416,3 +435,40 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(found).NotTo(BeFalse())
+		}
+	}
+
+	return endpointName, healthCheckIP
+}
+
+func (f *Framework) SetHealthCheckIp(cluster framework.ClusterIndex, ip, endpointName string) {
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:465]:
```suggestion
func (f *Framework) SetHealthCheckIP(cluster framework.ClusterIndex, ip, endpointName string) {
```

File: test/e2e/framework/framework.go [hunk 5]
```diff
@@ -416,3 +435,40 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+	By(fmt.Sprintf("Setting Health Check Ip to %v in cluster %q", ip, framework.TestContext.ClusterIDs[cluster]))
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:466]:
```suggestion
	By(fmt.Sprintf("Setting health check IP cluster %q to %v", framework.TestContext.ClusterIDs[cluster], ip))
```

File: test/e2e/framework/framework.go [hunk 6]
```diff
@@ -416,3 +435,40 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(found).NotTo(BeFalse())
+			healthCheckIP, found, err = unstructured.NestedString(spec, "healthCheckIP")
+
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(found).NotTo(BeFalse())
+		}
+	}
+
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:461]: This should use `framework.AwaitUntil` just in case.

File: test/e2e/discovery/service_discovery.go [hunk 7]
```diff
@@ -426,6 +432,67 @@ func RunServiceDiscoveryRoundRobinTest(f *lhframework.Framework) {
...
+
+	randomIP := "192.168.1.5"
+	endpointName, healthCheckIP := f.GetHealthCheckIpInfo(framework.ClusterC)
+	f.SetHealthCheckIp(framework.ClusterC, randomIP, endpointName)
+
+	verifyServiceIpWithDig(f.Framework, framework.ClusterA, framework.ClusterB, nginxServiceClusterB, netshootPodList,
+		checkedDomains, "", true)
+	verifyServiceIpWithDig(f.Framework, framework.ClusterA, framework.ClusterC, nginxServiceClusterC, netshootPodList,
+		checkedDomains, "", false)
+	f.SetHealthCheckIp(framework.ClusterC, healthCheckIP, endpointName)
```
Comment (tpantelis) [inline test/e2e/discovery/service_discovery.go:493]: This really should be done in `AfterEach` to ensure we leave it in a good state if the test fails. Call `GetHealthCheckIPInfo` in `BeforeEach`.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: test/e2e/framework/framework.go [hunk 8]
```diff
@@ -416,3 +435,39 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+		return unstructuredEndpointList, err
+	}, framework.NoopCheckResult)
+
+	unstructuredEndpointList := endpointList.(*unstructured.UnstructuredList)
+	for _, endpoint := range unstructuredEndpointList.Items {
+		By(fmt.Sprintf("Getting the endpoint %s, for cluster %s", endpoint.GetName(), framework.TestContext.ClusterIDs[cluster]))
+
+		if strings.Contains(endpoint.GetName(), framework.TestContext.ClusterIDs[cluster]) {
+			endpointName = endpoint.GetName()
+			Expect(endpointName).NotTo(BeNil())
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:451]: The Name can't be nil.

File: test/e2e/framework/framework.go [hunk 9]
```diff
@@ -416,3 +435,39 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+
+			var found bool
+			var err error
+			healthCheckIP, found, err = unstructured.NestedString(endpoint.Object, "spec", "healthCheckIP")
+
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(found).NotTo(BeFalse())
+		}
+	}
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:460]: All of this should be within the check result function of `AwaitUntil`.

File: test/e2e/discovery/service_discovery.go [hunk 10]
```diff
@@ -81,6 +81,21 @@ var _ = Describe("[discovery] Test Service Discovery Across Clusters", func() {
...
+			randomIP := "192.168.1.5"
+			endpointName, healthCheckIP = f.GetHealthCheckIPInfo(framework.ClusterC)
+			f.SetHealthCheckIP(framework.ClusterC, randomIP, endpointName)
+		})
+		It("should not resolve the service", func() {
+			RunServicesClusterAvailabilityMutliClusterTest(f)
+		})
+		AfterEach(func() {
+			f.SetHealthCheckIP(framework.ClusterC, healthCheckIP, endpointName)
+		})
```
Comment (tpantelis) [inline test/e2e/discovery/service_discovery.go:97]: Can you put newlines in between the functions (I'm surprised the linter didn't complain)? Scrunched code is less readable.

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: test/e2e/discovery/service_discovery.go [hunk 11]
```diff
@@ -426,6 +444,57 @@ func RunServiceDiscoveryRoundRobinTest(f *lhframework.Framework) {
 	verifyRoundRobinWithDig(f.Framework, framework.ClusterA, nginxServiceClusterB.Name, serviceIPList, netshootPodList, checkedDomains)
 }
 
+func RunServicesClusterAvailabilityMutliClusterTest(f *lhframework.Framework) {
+	clusterBName := framework.TestContext.ClusterIDs[framework.ClusterB]
+	clusterCName := framework.TestContext.ClusterIDs[framework.ClusterC]
```
Comment (vthapar) [inline test/e2e/discovery/service_discovery.go:449]: Check for number of clusters.

**Review by vthapar — CHANGES_REQUESTED** [review]

File: test/e2e/discovery/service_discovery.go (hunk 11 above)
Comment (aswinsuryan) [reply test/e2e/discovery/service_discovery.go:469]: done.

File: test/e2e/discovery/service_discovery.go [hunk 12]
```diff
@@ -81,6 +81,29 @@ var _ = Describe("[discovery] Test Service Discovery Across Clusters", func() {
...
+				Skip("Only two clusters are deployed and hence skipping the test")
+				return
+			}
+
+			randomIP := "192.168.1.5"
+			endpointName, healthCheckIP = f.GetHealthCheckIPInfo(framework.ClusterC)
+			f.SetHealthCheckIP(framework.ClusterC, randomIP, endpointName)
+		})
+
+		It("should not resolve the service", func() {
```
Comment (tpantelis) [inline test/e2e/discovery/service_discovery.go:99]:
```suggestion
		It("should not resolve that cluster's service IP", func() {
```

File: test/e2e/discovery/service_discovery.go [hunk 13]
```diff
@@ -81,6 +81,29 @@ var _ = Describe("[discovery] Test Service Discovery Across Clusters", func() {
 			RunServiceDiscoveryRoundRobinTest(f)
 		})
 	})
+
+	When("only one of the clusters with a service is healthy", func() {
```
Comment (tpantelis) [inline test/e2e/discovery/service_discovery.go:85]:
```suggestion
	When("one of the clusters with a service is not healthy", func() {
```

File: test/e2e/framework/framework.go [hunk 14]
```diff
@@ -416,3 +435,42 @@ func (f *Framework) SetNginxStatefulSetReplicas(cluster framework.ClusterIndex,
...
+
+				var found bool
+				var err error
+				healthCheckIP, found, err = unstructured.NestedString(endpoint.Object, "spec", "healthCheckIP")
+
+				if err != nil {
+					return false, fmt.Sprintf("Error retrieving the HealthcheckIP from spec : %v ", endpoint), nil
+				} else if !found {
+					return false, fmt.Sprintf("HealthcheckIP is not found in the spec : %v ", endpoint), nil
+				}
```
Comment (tpantelis) [inline test/e2e/framework/framework.go:459]:
```suggestion
				if err != nil {
					return false, "", err
				} 
				
				if !found {
					return false, fmt.Sprintf("HealthcheckIP not found in %#v ", endpoint), nil
				}
```

**Review by tpantelis — CHANGES_REQUESTED** [review]
Looks good - just a couple other suggestions.

**Comment by aswinsuryan** [comment]
Depends on https://github.com/submariner-io/submariner-charts/pull/106 , to fix the helm e2e failure.

**Review by tpantelis — APPROVED** [review]

**Review by vthapar — APPROVED** [review]

**Comment by dfarrell07** [comment]
This was backported in https://github.com/submariner-io/lighthouse/pull/450.

**Pull request merged**